demo_apps/llama-on-prem.md

```diff
@@ -0,0 +1,184 @@
+conda create -n vllm python=3.8
+conda activate vllm
+pip install vllm
+```
+
+Then run `huggingface-cli login` and copy and paste your Hugging Face access token to complete the login.
+
+There are two ways to deploy Llama 2 via vLLM, as a general API server or an OpenAI-compatible server (see [here](https://platform.openai.com/docs/api-reference/authentication) on how the OpenAI API authenticates, but you won't need to provide a real OpenAI API key when running Llama 2 via vLLM in the OpenAI-compatible mode).
+
+### Deploying Llama 2 as an API Server
+
+Run the command below to deploy vLLM as a general Llama 2 service:
+
+```
+python -m vllm.entrypoints.api_server --host 0.0.0.0 --port 5000 --model meta-llama/Llama-2-7b-chat-hf
+```
+
+Then on another terminal you can run:
+
+```
+curl http://localhost:5000/generate -d '{
+        "prompt": "Who wrote the book Innovators dilemma?",
+        "max_tokens": 300,
+        "temperature": 0
+    }'
+```
+
+to send a query (prompt) to Llama 2 via vLLM and get Llama's response:
+
+> Who wrote the book Innovators dilemma? The book "Innovator's Dilemma" was written by Clayton M. Christensen. It was first published in 1997 and has since become a classic in the field of business and innovation. In the book, Christensen argues that successful companies often struggle to adapt to disruptive technologies and new market entrants, and that this struggle can lead to their downfall. He also introduces the concept of the "innovator's dilemma," which refers to the paradoxical situation in which a company's efforts to improve its existing products or services can actually lead to its own decline.
+
+Now in your Llama client app, you can make an HTTP request as the `curl` command above to send a query to Llama and parse the response.
+
+If you add the port 5000 to your EC2 instance's security group's inbound rules with the TCP protocol, then you can run this on your Mac/Windows for test:
+
+```
+curl http://<EC2_public_ip>:5000/generate -d '{
+        "prompt": "Who wrote the book godfather?",
+        "max_tokens": 300,
+        "temperature": 0
+    }'
+```
+
+Also, if you have multiple GPUs, you can add the `--tensor-parallel-size` argument when starting the server (see [here](https://vllm.readthedocs.io/en/latest/serving/distributed_serving.html) for more info). For example, the command below runs the Llama 2 13b-chat model on 4 GPUs:
+
+```
+python api_server.py --host 0.0.0.0 --port 5000 --model meta-llama/Llama-2-13b-chat-hf --tensor-parallel-size 4
+```
+With multiple GPUs, you can also run replica of models as long as your model size can fit into targeted GPU memory. For example, if you have two A10G with 24 GB memory, you can run two 7B Llama 2 models at the same time. This can be done by launching two api servers each targeting specific CUDA cores on different ports:
+`CUDA_VISIBLE_DEVICES=0 python api_server.py --host 0.0.0.0 --port 5000  --model meta-llama/Llama-2-7b-chat-hf --tensor-parallel-size 1 `
```

Comment:
@jeffxtang I am not sure if we need `--tensor-parallel-size 1 ` here as thats the default.

Reply:
@HamidShojanazeri actually @WuhanMonkey Hector added this paragraph and looks like you're right - with CUDA_VISIBLE_DEVICES set to be a single value, --tensor-parallel-size shouldn't be needed. Hector can you please confirm?

Reply:
True. Let me change it